### DISCUSSION TITLE
Grain reconstruction from EBSD grainIds

### DISCUSSION THREAD

**Discussion by frankNiessen** [mtex-toolbox/mtex#1877] [Ask Anything] [answered]
Hi, 

assuming I have grainIds written into my EBSD object but I do not have the associated grain2D object,is it possible to efficiently generate this object with calcGrains.m or in another way?

I am asking since I have two EBSD data subsets for which I have used different methods for computing the grains and it is not per se possible to concatenate two grain2D objects. Maybe I can achieve the latter by constructing a grain2D object based on the ebsd.grainId information. 

Thanks for you help!

**Comment by nyyssont** [comment]
I think you could use the "variants" option in calcGrains - just give the grainId twice. It is a hack but might work anyway...

**Comment by frankNiessen** [comment]
Hi Tuomo, 

that was my initial idea, but the code turned out to run virtually infinitely, so I think it is extremely inefficient. You would think once the grainIDs for each EBSD pixel are known, it should be a quite fast procedure to get the grains. My best guess is that the grain2D object could be constructed directly without having to "calculate" the grains, but I am not sure it is possible.

**Comment by kilir** [comment, marked as answer]
Hi Frank,

while this is still not a  "direct" grain reconstruction -  somehow mtex needs to compute the grain vertices - it also works quite ok for data such as grainIds coming from e.g.  Aztecs processed data, which the backdraw that grain edges will always on a grid.  

```
mtexdata titanium

[grains1,ebsd.grainId]=ebsd.calcGrains
plot(grains1); nextAxis
%% compute grains from ID
grains2 = ebsd.calcGrains('custom',ebsd.grainId,0.5,'unitCell')
plot(grains2)
```


Cheers,
Rüdiger

**Reply by frankNiessen** [reply]
Thanks Ruediger, worked like a charm!

**Reply by frankNiessen** [reply]
Hi Rüdiger, 

maybe a quick follow-up question. I have some unindexed EBSD datapoints and when using your method on the indexed data I get some ugly single pixel areas where no grains are computed. If I on the other hand compute the grains for the entire ebsd dataset, including the unindexed data, and set the grainIds at these locations to NaN, I get some weird distorted grain map - this does not work. 

Is there an equivalent way of getting nice grains as with `calcGrains(ebsd('indexed'),'angle',3*degree)` using the custom method? I cannot think of how to achieve this, but it would be quite important.

Best wishes
Frank

**Reply by kilir** [reply]
Hi Frank,
if I understand correctly, simply skip the 'unitcell'.
Cheers,
Rüdiger

**Reply by frankNiessen** [reply]
Thanks. Now it is unfortunately just as slow as the "variants" option. Hmm I am lost :)

**Reply by frankNiessen** [reply]
... As a short update - runnings the `calcGrains('variants')` command is significantly sped up when running with gridified EBSD data (thanks @AzdiarGazder ). It still takes 5 minutes for my 1.5 Million Pixels dataset, but this is a lot faster than previously.
